Show CPU & memory usage in webui

swsnr opened the pull request:
For #287 

Shows the raw numbers currently.  Comparing resources against the reserved resources by the Nomad job description as suggested in #287 comes next, but not as part of this PR.

swsnr committed:
Use tagged type for docker handle

swsnr committed:
Add models for allocation resource usage

swsnr committed:
Make access to Node client API explicit

Do not implicitly query allocation nodes.  Require callers to explicitly
get hold of an allocation node to call client API on.

swsnr committed:
Test getInstanceTasks

swsnr committed:
Flatten allocated tasks in backend

Makes the result resemble the presentation in the UI

swsnr committed:
Fix client API URL

swsnr committed:
Add partial unification compiler plugin

Allows us to use .sequence over Either

swsnr committed:
Add type alias for instance results

swsnr committed:
Convert sequenced calls into for comprehension

swsnr committed:
Define missing reads instance

swsnr committed:
Include CPU ticks and memory usage in allocated tasks

swsnr committed:
Cache instance tasks for three seconds

Reduces load on Nomad when frontends request instance tasks once per second

swsnr committed:
Simplify type

swsnr committed:
Show current memory and CPU usage in UI

swsnr committed:
Fix unit tests for nomad instances

swsnr committed:
Fix elm tests

FRosner commented on webui/src/Views/InstanceView.elm:
```diff
@@ -441,23 +437,28 @@ instanceAllocationRow instance index ( taskName, allocation ) =
-            , td [] [ text taskName ]
+            , td [] [ text task.taskName ]
+            , td [ class "text-center" ]
+                [ text (unwrap "unknown" (Round.round 2) task.cpuTicksUsed)
```
We should also add the unit here (MHz).

swsnr replied:
I'm not sure these are MHz.  Nomad exposes them as CPU ticks, in my understanding that's not the same as MHz.

FRosner replied:
It corresponds to CPU shares (MHz) and we are going to put it into relation to the allocated CPU shares (MHz) in order to get the percentage.

```
Task Resources
CPU         Memory          Disk     IOPS  Addresses
33/100 MHz  18 MiB/128 MiB  300 MiB  0     ui: 127.0.0.1:8000
```

![image](https://user-images.githubusercontent.com/3427394/29869130-f7648396-8d81-11e7-8263-9da03104dde0.png)

- https://www.nomadproject.io/docs/agent/telemetry.html#host-metrics

swsnr replied:
@FRosner I was confused by the name “ticks“ to be honest.  In Unix "tick" normally refers to the internal timer of the kernel not to CPU clock speed.

FRosner commented on webui/webpack.config.js:
```diff
@@ -83,7 +83,7 @@ module.exports = {
         exclude: [/elm-stuff/, /node-modules/],
         use: [
-          { loader: 'elm-hot-loader' },
+          // { loader: 'elm-hot-loader' },
```
Why did you comment out the hot loader?

swsnr replied:
It broke with some elm module that I introduced.  I'm not sure why, but I wanted to take a look first.

FRosner commented on server/src/main/scala/de/frosner/broccoli/websocket/WebSocketMessageHandler.scala:
```diff
@@ -58,3 +63,53 @@ class BroccoliMessageHandler @Inject()(instances: NomadInstances, instanceServic
+)(implicit ec: ExecutionContext)
+    extends WebSocketMessageHandler {
+
+  private val timeout: Duration = 3.seconds
```
This is how I understand the mechanism: Instead of limiting the clients in the amount of requests they can make to Broccoli but then forward every request to nomad directly we are allowing the clients to make as many requests as they like, but then cache the nomad response for three seconds. Is this correct?

I think we should make the cache timeout configurable so we can find an optimal tradeoff between the up-to-dateness and the load we put on Nomad. What do you think?

swsnr replied:
That's the approach, yes.  I'll add a configuration value for the timeout.

swsnr committed:
Don't inject nomad configuration as singleton

Broccoli configuration is already a singleton

swsnr committed:
Remove unused local binding

swsnr committed:
Extract websocket cache timeout to configuration

swsnr committed:
Parse and show CPU ticks as MHz

swsnr committed:
Trigger elm-hot-loader before babel

elm-hot-loader doesn't play well with babelified JS; it needs the JS as returned by the Elm compiler.
Otherwise it complains: "Module build failed: Error: elm-hot-loader is incompatible with this version of Elm."

Triggering it before Babel transpilation fixes this issue and seems to preserve hot-reloading.

swsnr committed:
Fix compiler error in Elm tests

swsnr commented:
@FRosner Can you review again?  This part is ready for merge as far as I'm concerned.

FRosner commented:
Will do!
-- 
Sent from my phone. Please excuse my brevity.

FRosner commented on server/src/main/resources/reference.conf:
```diff
@@ -41,6 +41,12 @@ broccoli {
+  # Configuration for Broccoli's Websocket connection with the webui
+  web-socket {
+    # How long to cache responses to websocket requests in Broccoli
+    cache-timeout = 3 seconds
```
We could actually use this timeout also for caching the other requests (like service and job status) later on, right? So the good design would be to ask Nomad on demand (not every x seconds) for the job status and everything but then cache the results?

swsnr replied:
@FRosner Yes, that's what I'd do.

FRosner commented:
LGTM. Just one comment: Right now the column size for CPU and Memory is not fixed. If the metrics now change, the columns will become bigger and smaller. I don't think this is an issue, because as soon as we replace the numbers by fixed width bars we shouldn't have this problem anymore.
